### PR TITLE
[cluster-management] filter out workflows/jobs from cluster's resources

### PR DESCRIPTION
**Pull request by kangnanli**
ConfigGenerator use helixAdmin to get resources from the cluster, which will include both db resources, and workflows/jobs; this will break the configGenerator. Thus, we need exclude the wf/jobs from the resources retrieved from cluster's idealStates, and only pass db_resources for config Generator.

### CODE REVIEW

File: cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java [hunk 1]
```diff
@@ -237,4 +242,37 @@ private boolean updateDisabledHosts() {
...
+    List<String> db_resources = new ArrayList<>();
+
+    Map<String, WorkflowConfig> workflowConfigMap = (driver == null ? null : driver.getWorkflows());
+
+    if (driver != null && workflowConfigMap != null && !workflowConfigMap.isEmpty()) {
+      Set<String> workflows = workflowConfigMap.keySet();
+      for (String resource : resources) {
+        boolean isWfOrJob = false;
+        for (String workflow : workflows) {
+          if (!workflow.isEmpty() && resource.startsWith(workflow)) {
```
Comment (newpoo) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java:262]: How do we guarantee that no db resource will start with any workflow name?

Is there any function/tag/something in helix zk that can help us tell if a resource is a regular db resource or workflow resource directly?

Comment (kangnanli) [reply cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java:262]: We have planned to enforce duplication check (specifically, prefix duplication check) when new db resource or workflow addition. This can prevent no db resource and workflow/jobs will have same name. 

I can't find any existing method to retrieve db only resources from zk. I will ask in helix community to confirm this.

Comment (newpoo) [reply cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java:262]: Will replacing startswith() with equals() work?

The implementation of this function can be refactored/simplified a little bit.

Comment (kangnanli) [reply cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java:262]: I got some feedback from helix team, will re-implement the filtering.

File: cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java [hunk 2]
```diff
@@ -237,4 +239,27 @@ private boolean updateDisabledHosts() {
     disabledHosts = latestDisabledInstances;
     return true;
   }
+
+  /**
+   * filter out resources with "Task" state model (ie. workflows and jobs);
+   * only return db resources from ideal states
+   *
+   * db_resource and wf/job have to be unique, bcz wf/job is treated as "resource" in helix.
+   * uniqueness will be enforced when workflow/job/db_resource is added
```
Comment (newpoo) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java:248]: remove this two lines, as they are not relevant any more.

File: cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java [hunk 3]
```diff
@@ -237,4 +239,27 @@ private boolean updateDisabledHosts() {
...
+   */
+  public List<String> filterOutNoneDbResources(List<String> resources) {
```
Comment (newpoo) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java:250]: can we rename it to filterOutTaskResources() to make it more accurate to what the function does?

File: cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java [hunk 4]
```diff
@@ -237,4 +239,27 @@ private boolean updateDisabledHosts() {
...
+    List<String> db_resources = new ArrayList<>();
+    for (String res : resources) {
+      IdealState ideal = admin.getResourceIdealState(clusterName, res);
+      if (ideal != null) {
+          String stateMode = ideal.getStateModelDefRef();
+          if (stateMode != null && !stateMode.equals("Task")) {
+            db_resources.add(res);
+          }
+      }
+    }
```
Comment (newpoo) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java:261]: Can we do it the other way around. i.e., filter out resources that we are sure they are Tasks? This seems to be safer to existing clusters.

File: cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java [hunk 5]
```diff
@@ -237,4 +242,27 @@ private boolean updateDisabledHosts() {
...
+      String res = iter.next();
+      IdealState ideal = admin.getResourceIdealState(clusterName, res);
+      if (ideal != null) {
+        String stateMode = ideal.getStateModelDefRef();
+        if (stateMode != null && stateMode.equals("Task")) {
+          iter.remove();
+        }
+      } else {
+        LOG.error(
+            "Did not remove resource from sharp map generation, due to can't get ideal state for "
```
Comment (newpoo) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java:262]: shard map

File: cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java [hunk 6]
```diff
@@ -92,14 +95,16 @@ public void onConfigChange(List<InstanceConfig> configs, NotificationContext cha
 
   @Override
   @PreFetch(enabled = false)
-  public void onExternalViewChange(List<ExternalView> externalViewList, NotificationContext changeContext) {
+  public void onExternalViewChange(List<ExternalView> externalViewList,
```
Comment (newpoo) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java:98]: undo this change?

File: cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java [hunk 7]
```diff
@@ -39,12 +40,14 @@
...
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
 
 public class ConfigGenerator extends RoutingTableProvider implements CustomCodeCallbackHandler {
+
```
Comment (newpoo) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java:50]: undo this change?

File: cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java [hunk 8]
```diff
@@ -237,4 +242,27 @@ private boolean updateDisabledHosts() {
...
     return true;
   }
+
+  /**
+   * filter out resources with "Task" state model (ie. workflows and jobs);
+   * only keep db resources from ideal states
+   */
+  public void filterOutTaskResources(List<String> resources) {
+    HelixAdmin admin = helixManager.getClusterManagmentTool();
+    for (Iterator<String> iter = resources.iterator(); iter.hasNext(); ) {
```
Comment (newpoo) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/ConfigGenerator.java:252]: nit: move Iterator<String> iter = resources.iterator() outside the loop and use while() instead of for()?
